File: ts/packages/core/src/stores.ts
// The persistence ports (PROTOCOL.md §1, §2) and their in-memory defaults.
//
// These are mekik's own state - the transcript and the conversation registry.
// They are deliberately separate from ilmek's checkpointer: ilmek owns graph
// state and open interrupts; mekik owns what a client sees and replays. v1
// ships only in-memory implementations; the ports exist so a durable backend
// (Redis, Postgres) can drop in later without touching the engine.

import type { OutgoingFrame } from "./protocol.ts";
import { PERSISTENT_FRAME_TYPES } from "./protocol.ts";

/** A persistent server→client frame (carries `seq`). */
export type PersistentFrame = Extract<OutgoingFrame, { seq: number }>;

/**
 * The transcript port. Stores the persistent frames of a conversation in `seq`
 * order and replays the tail after a watermark. `seq` is assigned by the engine
 * (it owns the monotonic counter); the store only persists and ranges.
 */
export interface HistoryStore {
    /** Persist an already-seq-stamped persistent frame. */
    record(conversationId: string, frame: PersistentFrame): Promise<void>;
    /** Persistent frames with `seq > watermark`, ascending - the reconnect replay. */
    after(conversationId: string, watermark: number): Promise<PersistentFrame[]>;
    /** Highest `seq` stored for the conversation, or 0 - used to seed the engine's counter. */
    currentSeq(conversationId: string): Promise<number>;
}

export class InMemoryHistoryStore implements HistoryStore {
    private readonly byConversation = new Map<string, PersistentFrame[]>();

    async record(conversationId: string, frame: PersistentFrame): Promise<void> {
        if (!(PERSISTENT_FRAME_TYPES as readonly string[]).includes(frame.type)) {
            throw new Error(`refusing to record a transient ${frame.type} frame in the transcript`);
        }
        const list = this.byConversation.get(conversationId) ?? [];
        list.push(frame);
        this.byConversation.set(conversationId, list);
    }

    async after(conversationId: string, watermark: number): Promise<PersistentFrame[]> {
        const list = this.byConversation.get(conversationId) ?? [];
        // Stored in append order, which is seq order - but filter by seq rather
        // than slice, so a store that ever reorders stays correct.
        return list.filter((f) => f.seq > watermark);
    }

    async currentSeq(conversationId: string): Promise<number> {
        const list = this.byConversation.get(conversationId);
        return list && list.length > 0 ? list[list.length - 1]!.seq : 0;
    }
}

// ── conversations ─────────────────────────────────────────────────────────────

export interface ConversationRecord {
    conversationId: string;
    /** The user who owns the conversation. */
    userId: string;
    createdAt: number;
    meta: Record<string, unknown>;
}

/** The conversation registry - owner and metadata. Open interrupts live in ilmek, not here. */
export interface ConversationStore {
    get(conversationId: string): Promise<ConversationRecord | null>;
    create(record: ConversationRecord): Promise<void>;
}

export class InMemoryConversationStore implements ConversationStore {
    private readonly byId = new Map<string, ConversationRecord>();

    async get(conversationId: string): Promise<ConversationRecord | null> {
        return this.byId.get(conversationId) ?? null;
    }

    async create(record: ConversationRecord): Promise<void> {
        this.byId.set(record.conversationId, record);
    }
}

File: dotnet/src/Mekik.Core/Stores.cs
namespace Mekik;

/// <summary>A wire frame — mekik models frames as dictionaries for byte-exact parity (PROTOCOL.md §9).</summary>
using Frame = System.Collections.Generic.IReadOnlyDictionary<string, object?>;

/// <summary>
/// The transcript port (PROTOCOL.md §2). Stores the persistent frames of a
/// conversation in `seq` order and replays the tail after a watermark. `seq` is
/// assigned by the engine; the store only persists and ranges. Separate from
/// ilmek's checkpointer: ilmek owns graph state, mekik owns what a client sees.
/// </summary>
public interface IHistoryStore
{
    Task RecordAsync(string conversationId, Frame frame);
    Task<IReadOnlyList<Frame>> AfterAsync(string conversationId, long watermark);
    Task<long> CurrentSeqAsync(string conversationId);
}

public sealed class InMemoryHistoryStore : IHistoryStore
{
    private readonly Dictionary<string, List<Frame>> _byConversation = new();
    private readonly object _lock = new();

    private static long SeqOf(Frame frame) => frame.TryGetValue("seq", out var s) ? Convert.ToInt64(s) : 0;

    public Task RecordAsync(string conversationId, Frame frame)
    {
        if (frame.GetValueOrDefault("type") is not string t || !Protocol.PersistentFrameTypes.Contains(t))
            throw new InvalidOperationException($"refusing to record a transient {frame.GetValueOrDefault("type")} frame");
        lock (_lock)
        {
            if (!_byConversation.TryGetValue(conversationId, out var list))
                _byConversation[conversationId] = list = new List<Frame>();
            list.Add(frame);
        }
        return Task.CompletedTask;
    }

    public Task<IReadOnlyList<Frame>> AfterAsync(string conversationId, long watermark)
    {
        lock (_lock)
        {
            var list = _byConversation.GetValueOrDefault(conversationId) ?? new List<Frame>();
            return Task.FromResult<IReadOnlyList<Frame>>(list.Where(f => SeqOf(f) > watermark).ToList());
        }
    }

    public Task<long> CurrentSeqAsync(string conversationId)
    {
        lock (_lock)
        {
            var list = _byConversation.GetValueOrDefault(conversationId);
            return Task.FromResult(list is { Count: > 0 } ? SeqOf(list[^1]) : 0L);
        }
    }
}

// ── conversations ─────────────────────────────────────────────────────────────

public sealed record ConversationRecord(string ConversationId, string UserId, long CreatedAt, IReadOnlyDictionary<string, object?> Meta);

/// <summary>The conversation registry — owner and metadata. Open interrupts live in ilmek, not here.</summary>
public interface IConversationStore
{
    Task<ConversationRecord?> GetAsync(string conversationId);
    Task CreateAsync(ConversationRecord record);
}

public sealed class InMemoryConversationStore : IConversationStore
{
    private readonly Dictionary<string, ConversationRecord> _byId = new();
    private readonly object _lock = new();

    public Task<ConversationRecord?> GetAsync(string conversationId)
    {
        lock (_lock) return Task.FromResult(_byId.GetValueOrDefault(conversationId));
    }

    public Task CreateAsync(ConversationRecord record)
    {
        lock (_lock) _byId[record.ConversationId] = record;
        return Task.CompletedTask;
    }
}
